core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/delete/DeleteKeystoresProcessor.java

```diff
@@ -0,0 +1,143 @@
+  private final ArtifactSignerProvider signerProvider;
+
+  public DeleteKeystoresProcessor(
+      final KeystoreFileManager keystoreFileManager,
+      final Optional<SlashingProtection> slashingProtection,
+      final ArtifactSignerProvider signerProvider) {
+    this.keystoreFileManager = keystoreFileManager;
+    this.slashingProtection = slashingProtection;
+    this.signerProvider = signerProvider;
+  }
+
+  public DeleteKeystoresResponse process(final DeleteKeystoresRequestBody requestBody) {
+    // normalize incoming keys to delete
+    final List<String> pubkeysToDelete =
+        requestBody.getPubkeys().stream()
+            .map(IdentifierUtils::normaliseIdentifier)
+            .collect(Collectors.toList());
+
+    final List<String> keysToExport = new ArrayList<>();
+    // attempt to delete keys one by one, and return results with statuses
+    final List<DeleteKeystoreResult> results = processKeysToDelete(pubkeysToDelete, keysToExport);
+    // export slashing data for keys that were either 'deleted' or 'not_active'
+    final String slashingProtectionExport = exportSlashingProtectionData(results, keysToExport);
+    return new DeleteKeystoresResponse(results, slashingProtectionExport);
+  }
+
+  private List<DeleteKeystoreResult> processKeysToDelete(
+      List<String> pubkeysToDelete, List<String> keysToExport) {
+    final List<DeleteKeystoreResult> results = new ArrayList<>();
+    // process each incoming key individually
+    for (String pubkey : pubkeysToDelete) {
+      try {
+        final Optional<ArtifactSigner> signer = signerProvider.getSigner(pubkey);
+
+        // check that key is active
+        if (signer.isEmpty()) {
+          // if not active, check if we ever had this key registered in the slashing DB
+          final boolean wasRegistered =
+              slashingProtection
+                  .map(protection -> protection.isRegisteredValidator(Bytes.fromHexString(pubkey)))
+                  .orElse(false);
+
+          // if it was registered previously, return not_active and add to list of keys to export,
+          // otherwise not_found
+          if (wasRegistered) {
+            keysToExport.add(pubkey);
+            results.add(new DeleteKeystoreResult(DeleteKeystoreStatus.NOT_ACTIVE, ""));
+          } else {
+            results.add(new DeleteKeystoreResult(DeleteKeystoreStatus.NOT_FOUND, ""));
+          }
+          continue;
+        }
+
+        // Check that key is read only, if so return an error status
+        if (signer.get() instanceof BlsArtifactSigner
+            && ((BlsArtifactSigner) signer.get()).isReadOnlyKey()) {
+          results.add(
+              new DeleteKeystoreResult(
+                  DeleteKeystoreStatus.ERROR, "Unable to delete readonly key: " + pubkey));
+          continue;
+        }
+
+        // Remove active key from memory first, will stop any further signing with this key
+        signerProvider.removeSigner(pubkey).get();
```

Comment:
nit: If we want removeSigner to be synchronous (i.e. wait until it's definitely deleted) for all usages then is there any need for removeSigner to be implemented with an executerService?

Reply:
I think the same could be said for addSigner

Reply:
the reason why I kept it using the executorService is because that way I can guarantee that the add/remove happens on a single thread (the executor used there is a `Executors.newSingleThreadExecutor()`)

Reply:
ah, makes sense - thanks for the explanation!